Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs

```diff
@@ -0,0 +1,347 @@
+            }
+            
+            var codeSubmissionReceived = new CodeSubmissionReceived(submitCode);
+
+            context.Publish(codeSubmissionReceived);
+
+            var code = submitCode.Code;
+            if (IsCompleteSubmission(code))
+            {
+                context.Publish(new CompleteCodeSubmissionReceived(submitCode));
+            }
+            else
+            {
+                context.Publish(new IncompleteCodeSubmissionReceived(submitCode));
+            }
+
+            if (submitCode.SubmissionType == SubmissionType.Diagnose)
+            {
+                return;
+            }
+
+            Exception exception = null;
+            Collection<string> results = null;
+            if (!cancellationSource.IsCancellationRequested)
+            {
+                await _runspaceSemaphore.WaitAsync();
+                try
+                {
+                    using(var ps = System.Management.Automation.PowerShell.Create(_runspace))
+                    {
+                        RegisterPowerShellStreams(ps, context, submitCode);
+                        try
+                        {
+                            results = ps
+                                .AddScript(code)
+                                .AddCommand("Out-String")
+                                .InvokeAndClearCommands<string>();
+                        }
+                        catch (Exception e)
+                        {
+                            results = ps
+                                .AddCommand("Out-String")
+                                    .AddParameter("InputObject", new ErrorRecord(e, null, ErrorCategory.NotSpecified, null))
+                                .InvokeAndClearCommands<string>();
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    exception = e;
+                }
+                finally
+                {
+                    _runspaceSemaphore.Release();
+                }
+            }
+
+            if (!cancellationSource.IsCancellationRequested)
+            {
+                if (exception != null)
+                {
+                    string message = null;
+                    context.Publish(new CommandFailed(exception, submitCode, message));
+                }
+                else
+                {
+                    if (results != null)
+                    {
+
+                        var formattedValues = FormattedValue.FromObject(results.Single());
+                        context.Publish(
+                            new ReturnValueProduced(
+                                results,
+                                submitCode,
+                                formattedValues));
+                    }
+                }
+            }
+            else
+            {
+                context.Publish(new CommandFailed(null, submitCode, "Command cancelled"));
```

Comment:
It's preferable here to do `context.Fail`. Directly publishing these commands might be something we do away with.